### PR TITLE
[AMDGPU] Support wide register or subregister access when emitting s_singleuse_vdst instructions.

### PR DESCRIPTION
**Pull request by ScottEgerton**
Both single use producer and consumer instructions using wide/sub
registers are now correctly tracked and eligible for being marked as
single use.

### CODE REVIEW

File: llvm/test/CodeGen/AMDGPU/insert-singleuse-vdst.mir [hunk 1]
```diff
@@ -398,7 +398,6 @@ body: |
   ; CHECK-NEXT:   successors: %bb.1(0x80000000)
   ; CHECK-NEXT:   liveins: $sgpr0_sgpr1
   ; CHECK-NEXT: {{  $}}
-  ; CHECK-NEXT:   S_SINGLEUSE_VDST 1
```
Comment (jayfoad) [inline llvm/test/CodeGen/AMDGPU/insert-singleuse-vdst.mir:401]: Why did this change? I don't see any wide register or subregister accesses related to the def of vgpr0 in this test.

Comment (ScottEgerton) [reply llvm/test/CodeGen/AMDGPU/insert-singleuse-vdst.mir:401]: I should have mentioned this in the commit message. This test should not be marked as single use as the use is after an exec mask change. This is caused by the change in AMDGPUInsertSingleUseVDST.cpp on line 110 replacing the RegisterUseCount.erase(Reg) with RegisterUseCount[Unit] = 0. When an exec mask change is detected, all registers in RegisterUseCount are given a value of 2 uses to exclude them from being marked as single use producers (line 92). However, when the values were being erased, they were no longer being correctly excluded. Setting the use count to 0 and keeping them in RegisterUseCount prevents this.

Comment (jayfoad) [reply llvm/test/CodeGen/AMDGPU/insert-singleuse-vdst.mir:401]: Could you split that fix out into a separate PR please?

Comment (ScottEgerton) [reply llvm/test/CodeGen/AMDGPU/insert-singleuse-vdst.mir:401]: Sure, no problem.

File: llvm/lib/Target/AMDGPU/AMDGPUInsertSingleUseVDST.cpp [hunk 2]
```diff
@@ -64,13 +64,15 @@ class AMDGPUInsertSingleUseVDST : public MachineFunctionPass {
...
-      DenseMap<MCPhysReg, unsigned> RegisterUseCount; // TODO: MCRegUnits
+      DenseMap<MCRegUnit, unsigned> RegisterUseCount;
 
       // Handle boundaries at the end of basic block separately to avoid
       // false positives. If they are live at the end of a basic block then
       // assume it has more uses later on.
-      for (const auto &Liveouts : MBB.liveouts())
-        RegisterUseCount[Liveouts.PhysReg] = 2;
+      for (const auto &Liveout : MBB.liveouts()) {
+        for (const MCRegUnit &Unit : TRI->regunits(Liveout.PhysReg))
```
Comment (jayfoad) [inline llvm/lib/Target/AMDGPU/AMDGPUInsertSingleUseVDST.cpp:73]: Really this ought to use `MCRegUnitMaskIterator` to iterate over only the regunits of `Liveout.PhysReg` that are enabled in `Liveout.LaneMask`. See other uses of `MCRegUnitMaskIterator` in `lib/CodeGen/` for examples. Or you can just leave a TODO if you don't want to implement that now.

Comment (ScottEgerton) [reply llvm/lib/Target/AMDGPU/AMDGPUInsertSingleUseVDST.cpp]: I wasn't aware of `MCRegUnitMaskIterator`. I've added this in.

**Review by jayfoad — APPROVED** [review]
LGTM.

**Pull request merged**